### ISSUE TITLE
Switch pgwatch2 to pgbouncer errors

### ISSUE THREAD

**Issue by yinan8128** [cybertec-postgresql/pgwatch2#348]
Hi pgwatch2 team,

Once we switch pgwatch2 to use pgbouncer, have error in pgwatch2 service, if pgwatch2 support to use pgbouncer or not? Please help check, thanks

pgwatch2 version: 1.8.3
OS: CentOS Linux release 7.6.1810
pgbouncer: 1.14.0
postgresql: 11.4

**systemctl status pgwatch2 -l**

pgwatch2.service - PGWatch2
   Loaded: loaded (/etc/systemd/system/pgwatch2.service; enabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/pgwatch2.service.d
           └─override.conf, required.conf
   Active: active (running) since Mon 2021-02-08 12:56:38 UTC; 1 weeks 1 days ago
 Main PID: 7972 (pgwatch2-daemon)
   CGroup: /system.slice/pgwatch2.service
           └─7972 /usr/bin/pgwatch2-daemon --datastore=prometheus --prometheus-port=9199 --add-real-dbname=true --host=127.0.0.1 --port=6432 --dbname=postgres --password=

Feb 08 12:56:38 s98765y-patroni_cluster-876274700.acronis systemd[1]: Starting PGWatch2...
Feb 08 12:56:38 s98765y-patroni_cluster-876274700.acronis systemd[1]: Started PGWatch2.
Feb 13 00:51:18 s98765y-patroni_cluster-876274700.acronis pgwatch2[7972]: 2021/02/13 00:51:18 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Feb 13 04:47:19 s98765y-patroni_cluster-876274700.acronis pgwatch2[7972]: 2021/02/13 04:47:19 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Feb 15 08:53:38 s98765y-patroni_cluster-876274700.acronis pgwatch2[7972]: 2021/02/15 08:53:38 ERRO ReadMetricDefinitionMapFromPostgres: write tcp 127.0.0.1:47094->127.0.0.1:6432: write: broken pipe
Feb 15 08:53:38 s98765y-patroni_cluster-876274700.acronis pgwatch2[7972]: 2021/02/15 08:53:38 ERRO main: Could not refresh metric definitions: write tcp 127.0.0.1:47094->127.0.0.1:6432: write: broken pipe

**Switch pgwatch2 to use pgbouncer changes, change port 5432 to 6432 and restart pgwatch2 service:**
**(1). pgwatch2.monitored_db record**

postgres=# select md_unique_name, md_hostname, md_port, md_dbname, md_user, md_preset_config_name, md_dbtype, md_include_pattern, md_exclude_pattern, md_host_config from pgwatch2.monitored_db;
-[ RECORD 1 ]---------+---------------------------------------------------------------------------------------------------------------------
md_unique_name        | patroni
md_hostname           | 127.0.0.1
md_port               | 6432
md_dbname             |
md_user               | pgwatch2
md_preset_config_name | exhaustive
md_dbtype             | postgres-continuous-discovery
md_include_pattern    | .*
md_exclude_pattern    | (template0|template1)
md_host_config        | {"scope": "pgcluster11", "dcs_type": "consul", "namespace": "/service/", "dcs_endpoints": ["http://localhost:8500"]}

**(2). /etc/sysconfig/pgwatch2.conf**
`
OPTIONS='--datastore=prometheus --prometheus-port=9199 --add-real-dbname=true --host=127.0.0.1 --port=6432 --dbname=postgres --password=
`

**Comment by kmoppel** [comment]
Hi! This "unnamed prepared statement does not exist" error hints that maybe you're using some non-default pooling mode? Better to monitor all DB-s directly as it haven't been tested over pgBouncer. Will try to find time to test via Bouncer also though so I'll leave the ticket open...

**Comment by yinan8128** [comment]
Hi Kmoppel,

We are using pgbouncer in common pooling mode: pool_mode = transaction

**Comment by yinan8128** [comment]
@kmoppel 

We use Prometheus mode and pgwatch2 agents running on all database hosts separately, using Config DB for metrics definition.

For #359, we only config pgwatch2.monitored_db md_port to 6432 and remain pgwatch2 service port 5432

And the operations should be like this:
Operation 1: pgwatch2 reading the configuration and metric definitions.
Operation 2: exposing metrics over a port for scraping in case of Prometheus mode.

As #359 configuration, only Operation 2 using pgbouncer 6432 port, but now the pgwatch2 service error is `GetMonitoredDatabasesFromConfigDB` which belong to Operation 1, so #359 is really a new ticket different with #348, could you please reopen #359 and help on it please? Thanks

Mar 25 20:32:30 s34567y-patroni_cluster-932280702.acronis pgwatch2[12752]: 2021/03/25 20:32:30 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist

**Comment by yinan8128** [comment]
@kmoppel 

Thank you for your reply

Actually the error didn't show once pgwatch2 service start, you can check this ticket and #359 ticket log, only start to have error after several hours or after several days, could you please monitor the env for a longer time please?

As you comment on pgwatch2.monitored_db record, could you please double confirm below setting is correct?

select md_unique_name, md_hostname, md_port, md_dbname, md_user, md_preset_config_name, md_dbtype, md_include_pattern, md_exclude_pattern, md_host_config from pgwatch2.monitored_db;
-[ RECORD 1 ]---------+---------------------------------------------------------------------------------------------------------------------
md_unique_name | patroni
md_hostname |
md_port | 6432
md_dbname |
md_user | pgwatch2
md_preset_config_name | exhaustive
md_dbtype | postgres-continuous-discovery
md_include_pattern |
md_exclude_pattern |
md_host_config | {"scope": "pgcluster11", "dcs_type": "consul", "namespace": "/service/", "dcs_endpoints": ["http://localhost:8500"]}

**Comment by kmoppel** [comment]
Ah ok, thanks for the extra context. Got the impression that the error appeared immediately. Then I'm guessing it was just a common network interruption ("write: broken pipe") or something with the local sockets. Very hard to simulate test that on my end...

**Comment by yinan8128** [comment]
@kmoppel 

As using Prometheus mode and pgwatch2 agents running on all database hosts separately, so not network problem

And for #359, we only config pgwatch2.monitored_db md_port to 6432 and remain pgwatch2 service port 5432

Before pgwatch2.monitored_db md_port using 5432 all works well, only once config to 6432 have "ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist", this is not caused by network or local socket, could you please reopen #359 and help?
